### PR TITLE
add multi share separate mount tests

### PR DESCRIPTION
**Pull request by gabemontero**
also rework parameter passing in e2e's

### CODE REVIEW

**Comment by gabemontero** [comment]
Interesting, in the disruptive test failure the pull-secret contents is different than what I typically see in testing:

```
        I0217 04:10:06.911369       1 hostpath.go:370] mapBackingResourceToPod
        I0217 04:10:06.911564       1 hostpath.go:202] create/update file /var/lib/kubelet/pods/60ce2189-5f47-4b66-aaf7-2ee9bbad1208/volumes/kubernetes.io~csi/my-csi-volume/mount/secrets/openshift-config:pull-secret/service-ca.crt
        I0217 04:10:06.911669       1 hostpath.go:202] create/update file /var/lib/kubelet/pods/60ce2189-5f47-4b66-aaf7-2ee9bbad1208/volumes/kubernetes.io~csi/my-csi-volume/mount/secrets/openshift-config:pull-secret/token
        I0217 04:10:06.911766       1 hostpath.go:202] create/update file /var/lib/kubelet/pods/60ce2189-5f47-4b66-aaf7-2ee9bbad1208/volumes/kubernetes.io~csi/my-csi-volume/mount/secrets/openshift-config:pull-secret/ca.crt

```

I may have to inspect the secret/configmaps as part of determining which content to test for.

**Comment by gabemontero** [comment]
the old content is still there too ... hmm

```
        /data/configmaps:
        total 0
        drwxr-xr-x    3 root     root            60 Feb 17 04:10 .
        drwxrwxrwt    4 root     root            80 Feb 17 04:10 ..
        drwxr-xr-x    2 root     root            80 Feb 17 04:10 openshift-config:openshift-install
        
        /data/configmaps/openshift-config:openshift-install:
        total 8
        drwxr-xr-x    2 root     root            80 Feb 17 04:10 .
        drwxr-xr-x    3 root     root            60 Feb 17 04:10 ..
        -rw-r--r--    1 root     root           124 Feb 17 04:10 invoker
        -rw-r--r--    1 root     root            70 Feb 17 04:10 version

```

**Comment by gabemontero** [comment]
> the old content is still there too ... hmm
> 
> ```
>         /data/configmaps:
>         total 0
>         drwxr-xr-x    3 root     root            60 Feb 17 04:10 .
>         drwxrwxrwt    4 root     root            80 Feb 17 04:10 ..
>         drwxr-xr-x    2 root     root            80 Feb 17 04:10 openshift-config:openshift-install
>         
>         /data/configmaps/openshift-config:openshift-install:
>         total 8
>         drwxr-xr-x    2 root     root            80 Feb 17 04:10 .
>         drwxr-xr-x    3 root     root            60 Feb 17 04:10 ..
>         -rw-r--r--    1 root     root           124 Feb 17 04:10 invoker
>         -rw-r--r--    1 root     root            70 Feb 17 04:10 version
> ```

the debug tracing provided enough light ... another juicy timing issue from our disruptive tests:

- the share add event for the restart and the share update event from our test came very close together (albeit the add came first)

 I0217 04:10:06.897081       1 shares.go:61] AddShare key openshift-config:openshift-install kind ConfigMap
 I0217 04:10:06.897209       1 shares.go:95] UpdateShare key e2e-test-csi-driver-projected-resource-pndz7 kind Secret

- but the way the thread scheduling went, the update share event completed processing before the add event

this left our dual state ... at least for the time being.  Enough for the test to complain

We would have been "eventually consistent" in that on the relist (default 10 minutes) we would have corrected things, but the test does not wait that long (nor do I think I want it to wait that long).

I think this reconciles with the pattern employed in the OCM build controller, where IDs are propagated as far as possible and objects are retrieved from the lister when payload is needed.  This would have at least narrowed the timing gap.

Right now, unless the test with this PR fails often enough to prevent merge, I want to fix it in a subsequent PR.

**Comment by gabemontero** [comment]
> > the old content is still there too ... hmm
> > ```
> >         /data/configmaps:
> >         total 0
> >         drwxr-xr-x    3 root     root            60 Feb 17 04:10 .
> >         drwxrwxrwt    4 root     root            80 Feb 17 04:10 ..
> >         drwxr-xr-x    2 root     root            80 Feb 17 04:10 openshift-config:openshift-install
> >         
> >         /data/configmaps/openshift-config:openshift-install:
> >         total 8
> >         drwxr-xr-x    2 root     root            80 Feb 17 04:10 .
> >         drwxr-xr-x    3 root     root            60 Feb 17 04:10 ..
> >         -rw-r--r--    1 root     root           124 Feb 17 04:10 invoker
> >         -rw-r--r--    1 root     root            70 Feb 17 04:10 version
> > ```
> 
> the debug tracing provided enough light ... another juicy timing issue from our disruptive tests:
> 
>     * the share add event for the restart and the share update event from our test came very close together (albeit the add came first)
> 
> 
> I0217 04:10:06.897081 1 shares.go:61] AddShare key openshift-config:openshift-install kind ConfigMap
> I0217 04:10:06.897209 1 shares.go:95] UpdateShare key e2e-test-csi-driver-projected-resource-pndz7 kind Secret
> 
>     * but the way the thread scheduling went, the update share event completed processing before the add event
> 
> 
> this left our dual state ... at least for the time being. Enough for the test to complain
> 
> We would have been "eventually consistent" in that on the relist (default 10 minutes) we would have corrected things, but the test does not wait that long (nor do I think I want it to wait that long).

Though changing the wait time to allow for eventual consistency is a "back up" option 

> 
> I think this reconciles with the pattern employed in the OCM build controller, where IDs are propagated as far as possible and objects are retrieved from the lister when payload is needed. This would have at least narrowed the timing gap.
> 
> Right now, unless the test with this PR fails often enough to prevent merge, I want to fix it in a subsequent PR.

File: test/e2e/disruptive_test.go [hunk 1]
```diff
@@ -9,17 +9,27 @@ import (
 )
 
 func TestBasicThenDriverRestartThenChangeShare(t *testing.T) {
```
Comment (otaviof) [inline test/e2e/disruptive_test.go:11]: As a suggestion for next interaction, `TestBasicThenDriverRestartThenChangeShare` could become a "test table", as we have when generating unit-tests.

File: test/e2e/slow_test.go [hunk 2]
```diff
@@ -10,16 +10,28 @@ import (
...
+	framework.CreateTestNamespace(testArgs)
+	defer framework.CleanupTestNamespace(testArgs)
+	basicShareSetupAndVerification(testArgs)
 
-	framework.DeleteShareRelatedRBAC(testNS, t)
-	t.Logf("%s: wait up to 10 minutes for examining pod %s since the controller does not currently watch all clusterroles and clusterrolebindings and reverse engineer which ones satisfied the SAR calls, so we wait for relist on shares", time.Now().String(), testNS)
-	framework.ExecPod(testNS, "openshift-config:openshift-install", true, 10*time.Minute, t)
+	framework.DeleteShareRelatedRBAC(testArgs)
+	t.Logf("%s: wait up to 10 minutes for examining pod %s since the controller does not currently watch all clusterroles and clusterrolebindings and reverse engineer which ones satisfied the SAR calls, so we wait for relist on shares", time.Now().String(), testArgs.Name)
+	testArgs.SearchStringMissing = true
```
Comment (otaviof) [inline test/e2e/slow_test.go:23]: Should we extract the interactions with `testArgs` to a function?

**Review by otaviof — COMMENTED** [review]
Good improvements to the test suite, @gabemontero. I've left a couple of minor comments for the next interactions. Kudos!

File: test/e2e/disruptive_test.go (hunk 1 above)
Comment (gabemontero) [reply test/e2e/disruptive_test.go:11]: yeah there has been some back and forth within our team @otaviof over that last year or so on the table approach vs. the explicit test case approach, both with unit and e2e tests, in the various repo's we maintain.

based on that prior history, I am at least of the mind that either approach is fine and we can allow the code writer a little freedom  / flexibility

running tests in parallel technically can come for free when using the test per method approach

there are means of running tests in parallel with the table approach as well, but you have to opt into the helpers etc. that facilitate that

File: test/e2e/slow_test.go (hunk 2 above)
Comment (gabemontero) [reply test/e2e/slow_test.go:23]: and make those fields themselves private ... yeah, I would agree that is more "academically correct" :-)

I've got a couple of other changes I have queued up that I think I'll refrain from further cluttering with that change, but taking it on afterward I think makes sense 

Or if you have a couple of cycles later next week and want to dabble in this repo a little :-) 

Thanks for the review / tag for merge !!

**Pull request merged**